File: JavaScript/Middle_of_the_Linked_List_876.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

/**
 * Method: use fast and slow node to visit the list. fast = fast.next.next; slow = slow.next
 *       When the fast == null or fast.next == null, then the slow is the node that need to find.
 */
var middleNode = function(head) {
    if (head == null || head.next == null) return head;

    let slow = head;
    let fast = head;

    while (fast != null && fast.next != null) {
        fast = fast.next.next;
        slow = slow.next;
    }

    return slow;
};

File: Java/Middle_of_the_Linked_List_876.java
package leetcode.Java;

/*
Given the head of a singly linked list, return the middle node of the linked list.
If there are two middle nodes, return the second middle node.
 */
public class Middle_of_the_Linked_List_876 {
    /*
      使用快慢指针
    */
    public ListNode middleNode(ListNode head) {
        if (head == null) return null;

        ListNode slow = head;
        ListNode fast = head;

        //不管是fast == null 还是 fast.next == null，循环停止时的slow指针即为所要找的中间指针
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }
}
